File: objective-c/tutorial3/tutorial3/ViewController.m
#import "ViewController.h"
@import RMQClient;

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self receiveLogs];
    [self receiveLogs];
    sleep(1);
    [self emitLog];
}

- (void)emitLog {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];
    RMQExchange *x = [ch fanout:@"logs"];

    NSString *msg = @"Hello World!";

    [x publish:msg];
    NSLog(@"Sent %@", msg);

    [conn close];
}

- (void)receiveLogs {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];
    RMQExchange *x = [ch fanout:@"logs"];
    RMQQueue *q = [ch queue:@"" options:RMQQueueDeclareExclusive];

    [q bind:x];

    NSLog(@"Waiting for logs.");

    [q subscribe:^(RMQMessage * _Nonnull message) {
        NSLog(@"Received %@", message);
    }];
}

@end

File: swift/tutorial3/tutorial3/ViewController.swift
//
//  ViewController.swift
//  tutorial3
//
//  Copyright © 2016 RabbitMQ. All rights reserved.
//
import UIKit

import RMQClient

class ViewController: UIViewController {


    override func viewDidLoad() {
        super.viewDidLoad()
        self.receiveLogs()
        self.receiveLogs()
        sleep(1)
        self.emitLog()
    }

    func emitLog() {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let x = ch.fanout("logs")
        let msg = "Hello World!"
        x.publish(msg.data(using: String.Encoding.utf8))
        print("Sent \(msg)")
        conn.close()
    }

    func receiveLogs() {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let x = ch.fanout("logs")
        let q = ch.queue("", options: .exclusive)
        q.bind(x)
        print("Waiting for logs.")
        q.subscribe({(_ message: RMQMessage) -> Void in
            print("Received \(String(data: message.body, encoding: String.Encoding.utf8)!)")
        })
    }
}
